interpreter/text/parser.mly

```diff
@@ -39,6 +39,45 @@ let ati i =
 let literal f s =
   try f s with Failure _ -> error s.at "constant out of range"
 
+let range_check i min max at =
+    let i_32 = Int32.to_int i in
+    let min_32 = Int32.of_int min in
+    let max_32 = Int32.of_int max in
+    if i > max_32 || i < min_32 then error at "constant out of range" else i_32
+
+let int8_of_string s = range_check (I32.of_string s.it) (-128) 255 s.at
+let int16_of_string s = range_check (I32.of_string s.it) (-32768) 65535 s.at
+let int32_of_string s = I32.to_bits (literal (fun s -> I32.of_string s.it) s)
+let int64_of_string s = I64.to_bits (literal (fun s -> I64.of_string s.it) s)
+let f32_of_string s = F32.to_bits (literal (fun s -> F32.of_string s.it) s)
+let f64_of_string s = F64.to_bits (literal (fun s -> F64.of_string s.it) s)
+
+let simd_literal f shape ss =
+    let open Bytes in
+    let b = create 16 in
+    (match shape with
+    | "i8x16" ->
+            List.iteri (fun i s -> set_uint8 b i (int8_of_string s)) ss;
```

Comment:
I can't figure out where these `set_` functions are coming from. But it would be nicer to abstract away the construction of a SIMD value from its lanes into some suitable constructor functions in the Values module.

Reply:
These come from Bytes (see `let open Bytes` right above).
"abstract away the construction of a SIMD value from its lanes"
I don't understand this, can you illustrate with an example? Since the "lanes" of a SIMD value (v128 value in the interpreter) are all the same (Bytes type), I'm not sure how having different constructors will help.
Do you mean having constructors like:
build_i8x16 : string -> V128.t
build_i16x8 : string -> V128.t
etc?

Reply:
> These come from Bytes (see `let open Bytes` right above).

This requires OCaml 4.08 (Ubuntu has 4.05 currently).

From source point of view lanes are not always bytes - `i16x8` has two byte lanes, `i32x4` -  four-bye and so on. I think passing array or list of strings, containing source text for corresponding lanes, would work for constructors.

Reply:
Yes you're right, I will add a note on OCaml 4.08 to the PR.